python-checks/src/main/java/org/sonar/python/checks/IdentityComparisonWithCachedTypesCheck.java

```diff
@@ -147,7 +148,7 @@ private boolean isUnsuitableOperand(Expression expr) {
   private boolean isUnsuitableType(PythonType type) {
     for (String builtinName : NAMES_OF_TYPES_UNSUITABLE_FOR_COMPARISON) {
       TypeCheckBuilder builtinWithNameChecker = typeChecker.typeCheckBuilder().isBuiltinWithName(builtinName);
-      if (builtinWithNameChecker.check(type) == TriBool.TRUE) {
+      if (type instanceof ObjectType && builtinWithNameChecker.check(type) == TriBool.TRUE) {
```

Comment:
I feel that we shouldn't mix checks done by the type checker and by the rule itself.
In this case, you want to ensure the object is actually an object of a specific type.

The proper fix (that you can consider in the type checking API sprint), is in my opinion the possibility to check something like `isObjectOfOf(PythonType expected)` that checks that we have an object of a certain type, rather than the type itself. I feel it's something that will come up enough that it will deserve its own method to create the relevant predicates.

As for the immediate fix, I think you could introduce a `isObject` method in the type checker that perform the same check. Then, thanks to the natural ANDing of predicates, you'll be able to write:
```
TypeCheckBuilder isObjectOfBuiltinCheck = typeChecker.typeCheckBuilder().isObject().isBuiltinWithName(builtinName);
```

And simply check that for TRUE.